js/web/lib/wasm/jsep/backend-webgpu.ts

```diff
@@ -0,0 +1,345 @@
+     inputGpuDataTypes: readonly GpuDataType[]): string => {
+      const inputTensorShapesToString = inputTensorShapes.map(d => `${d.join(',')}`).join('_');
+      const inputGpuDataTypesToString = inputGpuDataTypes.join('_');
+      let key = programInfo.name;
+      if (programInfo.cacheHint) {
+        key += '[' + programInfo.cacheHint + ']';
+      }
+      key += ':' + inputTensorShapesToString + ';' + inputGpuDataTypesToString;
+      return key;
+    };
+
+/**
+ * this class is designed to store status and being used as a singleton for JSEP. It will be passed to jsepInit() as
+ * the first parameter so that it is stored for future use.
+ */
+export class WebGpuBackend {
+  device: GPUDevice;
+  /**
+   * an instance of GpuDataManager to manage a GpuDataId -> GpuBuffer mapping
+   */
+  gpuDataManager: GpuDataManager;
+  /**
+   * an instance of ProgramManager to build and run WebGPU compute shader program, and manage a ProgramKey -> Program
+   * artifacts mapping
+   */
+  programManager: ProgramManager;
+
+  /**
+   * representing the kernel ID of which is currently being computed (CPU code perspective).
+   * `null` means no kernel is being computed.
+   * only one kernel can be computed at a moment.
+   */
+  currentKernelId: number|null = null;
+  /**
+   * a list of temporary GPU data for the current kernel. should release when the kernel done computation.
+   */
+  private temporaryData: GpuData[];
+  /**
+   * a KernelID -> a GPU data list, which stores persistent GPU data owned by the specific kernel.
+   */
+  private kernelPersistentData: Map<number, GpuData[]>;
+  /**
+   * a KernelID -> a custom data, which stores custom data owned by the specific kernel.
+   */
+  private kernelCustomData: Map<number, {[key: string]: unknown}>;
+  /**
+   * get the custom data of the current kernel
+   */
+  get currentKernelCustomData(): {[key: string]: unknown} {
+    if (this.currentKernelId === null) {
+      throw new Error('currentKernelCustomData(): currentKernelId is null. (should not happen)');
+    }
+
+    let data = this.kernelCustomData.get(this.currentKernelId);
+    if (!data) {
+      data = {};
+      this.kernelCustomData.set(this.currentKernelId, data);
+    }
+
+    return data;
+  }
+
+  /**
+   * a KernelID -> kernel info mapping. value is [ name, run function, [optional] preprocess_attribute_once function ]
+   */
+  kernels: Map<number, [string, RunFunction, [((attribute: unknown) => unknown) | undefined, unknown]]>;
+
+  commandEncoder: GPUCommandEncoder|null = null;
+  computePassEncoder: GPUComputePassEncoder|null = null;
+  pendingDispatchNumber = 0;
+
+  profilingEnabled = false;
+  profilingQuerySet: GPUQuerySet;
+  profilingTimeBase?: bigint;
+
+  async initialize(): Promise<void> {
+    if (!navigator.gpu) {
+      // WebGPU is not available.
+      throw new Error('WebGpuBackend: WebGPU is not available.');
+    }
+
+    const adapter = await navigator.gpu.requestAdapter();
```

Comment:
May I recommend passing [`powerPreference`](https://gpuweb.github.io/gpuweb/#dom-gpurequestadapteroptions-powerpreference) when requesting a GPU adapter so that developers can request which type of GPU they're looking for.

In https://github.com/xenova/transformers.js/pull/545 for instance, it would be preferable to test the "high-performance" GPU.
 
```suggestion
    const adapter = await navigator.gpu.requestAdapter({
      powerPreference: 'high-performance'
    });
```

Reply:
That would be great! 🔥  It will also be useful if we can get the selected adapter, without having to re-request an adapter. cc @guschmue

Reply:
Thanks for the suggestion. Will think about this.

Reply:
#19857 is created to address this. Please take a look

Reply:
Per my understanding on some OSes, like Windows, if we have multi-GPU and integrated GPU is the one chosen by Chrome during startup, simply set powerPreference to high-performance will not force WebGPU to utilize the discrete GPU.

Reply:
That's currently true on Windows. See https://source.chromium.org/chromium/chromium/src/+/main:gpu/command_buffer/service/webgpu_decoder_impl.cc;drc=5bc3326c91b582cbec543bc9896201a1c56bdebd;l=1621
On macOS, this is not the case.

Reply:
FYI for Windows, here's the bug: https://issues.chromium.org/issues/329211593